bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java

```diff
@@ -0,0 +1,537 @@
+ * {@link CompletableFuture}s.
+ * Each instance has a reference to the last CompletableFuture and will call handleAsync to add a new task.
+ *
+ * @author Jörg Sautter - Initial contribution
+ */
+class SequentialScheduledExecutorService implements ScheduledExecutorService {
+
+    private final WorkQueueEntry empty;
+    private final ScheduledThreadPoolExecutor pool;
+    private final List<RunnableFuture<?>> scheduled;
+    private final ScheduledFuture<?> cleaner;
+    private WorkQueueEntry tail;
+
+    public SequentialScheduledExecutorService(ScheduledThreadPoolExecutor pool) {
+        if (pool.getMaximumPoolSize() != Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("the pool must scale unlimited to avoid potential dead locks!");
+        }
+
+        this.pool = pool;
+
+        // prepare the WorkQueueEntry we are using when no tasks are pending
+        RunnableCompletableFuture<?> future = new RunnableCompletableFuture();
+        future.complete(null);
+        empty = new WorkQueueEntry(null, null, future);
+
+        // tracks scheduled tasks alive
+        this.scheduled = new ArrayList<>();
+
+        tail = empty;
+
+        // clean up to ensure we do not keep references to old tasks
+        cleaner = this.scheduleAtFixedRate(() -> {
+            synchronized (this) {
+                scheduled.removeIf((sf) -> sf.isCancelled());
+
+                if (tail == null) {
+                    // the service is shutdown
+                    return;
+                }
+
+                WorkQueueEntry entry = tail;
+
+                while (entry.prev != null) {
+                    if (entry.prev.future.isDone()) {
+                        entry.prev = null;
+                        break;
+                    }
+                    entry = entry.prev;
+                }
+
+                if (tail != empty && tail.future.isDone()) {
+                    // replace the tail with empty to ensure we do not prevent GC
+                    tail = empty;
+                }
+            }
+        }, 2, 4, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+        synchronized (this) {
+            if (tail == null) {
+                throw new RejectedExecutionException("this scheduled executor has been shutdown before");
+            }
+
+            CompletableFuture<RunnableFuture<?>> origin = new CompletableFuture<>();
+            ScheduledFuture<?> future = pool.schedule(() -> {
+                // we block the thread here, in worst case new threads are spawned
+                submit0(origin.join(), command).join();
+            }, delay, unit);
+
+            scheduled.add((RunnableFuture<?>) future);
+            origin.complete((RunnableFuture<?>) future);
+
+            return future;
+        }
+    }
+
+    @Override
+    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
```

Comment:
Can we somehow avoid the code duplication here (and in the next two methods)?

Reply:
Looks like this is possible, will fix this.